crates/shared/src/utils/mod.rs

```diff
@@ -0,0 +1,115 @@
+use std::{future::Future, hash::Hash, sync::Arc};
+
+use futures::{
+    future::{RemoteHandle, Shared},
+    FutureExt,
+};
+use hotshot::traits::BlockPayload;
+use hotshot_types::{
+    data::{DaProposal, QuorumProposal},
+    traits::{
+        block_contents::BlockHeader, node_implementation::NodeType,
+        signature_key::BuilderSignatureKey,
+    },
+    utils::BuilderCommitment,
+};
+use tokio::{spawn, sync::Notify};
+
+pub mod rotating_set;
+pub use rotating_set::RotatingSet;
+
+pub mod event_serivce_wrapper;
+pub use event_serivce_wrapper::EventServiceStream;
+
+/// A convenience type alias for a tuple of builder keys
+/// `(public_key, private_key)`
+pub type BuilderKeys<Types> = (
+    <Types as NodeType>::BuilderSignatureKey,
+    <<Types as NodeType>::BuilderSignatureKey as BuilderSignatureKey>::BuilderPrivateKey,
+);
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct ProposalId<Types>
+where
+    Types: NodeType,
+{
+    view_number: Types::View,
+    builder_commitment: BuilderCommitment,
+}
+
+impl<Types> ProposalId<Types>
+where
+    Types: NodeType,
+{
+    pub fn from_quorum_proposal(proposal: &QuorumProposal<Types>) -> Self {
+        Self {
+            builder_commitment: proposal.block_header.builder_commitment(),
+            view_number: proposal.view_number,
+        }
+    }
+
+    pub fn from_da_proposal(proposal: &DaProposal<Types>) -> Self {
+        let builder_commitment = <Types::BlockPayload as BlockPayload<Types>>::from_bytes(
+            &proposal.encoded_transactions,
+            &proposal.metadata,
+        )
+        .builder_commitment(&proposal.metadata);
+
+        Self {
+            builder_commitment,
+            view_number: proposal.view_number,
+        }
+    }
+
+    pub fn view_number(&self) -> <Types as NodeType>::View {
+        self.view_number
+    }
+
+    pub fn builder_commitment(&self) -> &BuilderCommitment {
+        &self.builder_commitment
+    }
+}
+
+#[derive(derive_more::Debug, Clone)]
+pub struct WaitAndKeep<T>
```

Comment:
This is changed quite a bit, but the general idea is essentially the same as the old [`WakeAndKeep`](https://github.com/EspressoSystems/marketplace-builder-core/blob/178a48f80d8bfd15a436c7697b3096ff9967538a/crates/legacy/src/lib.rs#L31): we keep around a handle to a future that is paused and we can
- activate it
- get result of it's execution